native/cocos/renderer/pipeline/custom/GraphView.h

```diff
@@ -0,0 +1,211 @@
+    typename AddressableView<GraphT>::vertex_descriptor u,
+    const AddressableView<GraphT>& g) noexcept {
+    return num_children(u, g.mGraph);
+}
+
+template <class GraphT>
+inline std::pair<typename AddressableView<GraphT>::edge_descriptor, bool>
+edge(typename AddressableView<GraphT>::vertex_descriptor u,
+     typename AddressableView<GraphT>::vertex_descriptor v,
+     const AddressableView<GraphT>& g) noexcept {
+    return ownership(u, v, g.mGraph);
+}
+
+// BidirectionalGraph
+template <class GraphT>
+inline std::pair<typename AddressableView<GraphT>::in_edge_iterator, typename AddressableView<GraphT>::in_edge_iterator>
+in_edges( // NOLINT(readability-identifier-naming)
+    typename AddressableView<GraphT>::vertex_descriptor u,
+    const AddressableView<GraphT>& g) noexcept {
+    return parents(u, g.mGraph);
+}
+
+template <class GraphT>
+inline typename AddressableView<GraphT>::degree_size_type
+in_degree( // NOLINT(readability-identifier-naming)
+    typename AddressableView<GraphT>::vertex_descriptor u,
+    const AddressableView<GraphT>& g) noexcept {
+    return num_parents(u, g.mGraph);
+}
+
+template <class GraphT>
+inline typename AddressableView<GraphT>::degree_size_type
+degree(typename AddressableView<GraphT>::vertex_descriptor u, const AddressableView<GraphT>& g) noexcept {
+    return out_degree(u, g) + in_degree(u, g);
+}
+
+// AdjacencyGraph
+template <class GraphT>
+inline std::pair<
+    typename AddressableView<GraphT>::adjacency_iterator,
+    typename AddressableView<GraphT>::adjacency_iterator>
+adjacent_vertices( // NOLINT(readability-identifier-naming)
+    typename AddressableView<GraphT>::vertex_descriptor u,
+    const AddressableView<GraphT>& g) noexcept {
+    auto r = out_edges(u, g);
+    return std::make_pair(
+        typename AddressableView<GraphT>::adjacency_iterator(r.first, &g),
+        typename AddressableView<GraphT>::adjacency_iterator(r.second, &g));
+}
+
+// VertexListGraph
+template <class GraphT>
+inline std::pair<
+    typename AddressableView<GraphT>::vertex_iterator,
+    typename AddressableView<GraphT>::vertex_iterator>
+vertices(const AddressableView<GraphT>& g) noexcept {
+    return vertices(g.mGraph);
+}
+
+template <class GraphT>
+inline typename AddressableView<GraphT>::vertices_size_type
+num_vertices(const AddressableView<GraphT>& g) noexcept { // NOLINT(readability-identifier-naming)
+    return num_vertices(g.mGraph);
+}
+
+// EdgeListGraph
+template <class GraphT>
+inline std::pair<
+    typename AddressableView<GraphT>::edge_iterator,
+    typename AddressableView<GraphT>::edge_iterator>
+edges(const AddressableView<GraphT>& g) noexcept {
+    return ownerships(g.mGraph);
+}
+
+template <class GraphT>
+inline typename AddressableView<GraphT>::edges_size_type
+num_edges(const AddressableView<GraphT>& g) noexcept { // NOLINT(readability-identifier-naming)
+    return num_ownerships(g.mGraph);
+}
+
+} // namespace render
+
+} // namespace cc
+
+namespace boost {
```

Comment:
Why implement it in boost namespace?

Reply:
It is specialization of boost::property_map, it is used by boost::graph.

Reply:
got it.